File: eval/corpora/pyapp/utils/retry.py
"""Retry helpers for calls to flaky external services (payment processors, carriers)."""

import random
import time


def retry_with_backoff(fn, attempts=3, base_delay=0.1):
    """Call `fn` up to `attempts` times, waiting an exponentially growing,
    jittered delay between failures so a burst of retries doesn't itself
    become a thundering herd against the downstream service.
    """
    last_exc = None
    for attempt in range(attempts):
        try:
            return fn()
        except Exception as exc:  # noqa: BLE001 - deliberately broad, generic retry wrapper
            last_exc = exc
            delay = base_delay * (2**attempt) + random.uniform(0, base_delay)
            time.sleep(delay)
    raise last_exc

File: eval/corpora/polyglot/service/src/utils/retry.rs
//! Retry helper for calls to flaky downstream services.

use std::thread::sleep;
use std::time::Duration;

/// Call `f` once, sleeping briefly on failure.
///
/// Production code would retry multiple times with jittered exponential
/// backoff so a burst of retries doesn't become its own thundering herd
/// against the downstream service; this fixture keeps it to a single
/// attempt so indexing stays deterministic.
pub fn retry_with_backoff<F: FnOnce() -> bool>(f: F) -> bool {
    let ok = f();
    if !ok {
        sleep(Duration::from_millis(1));
    }
    ok
}
